File: day4/day4_2.cpp
#include <fstream>
#include <string>
#include <iostream>
#include <vector>
#include <array>


inline bool check_vals(char x, char y) {
    return (x == 'M' && y == 'S') || (x == 'S' && y == 'M');
}

bool check_coord(std::vector<std::string> const& input, int row_num, int col_num) {
    if (input[row_num][col_num] != 'A') {
        return false;
    }

    return check_vals(input[row_num - 1][col_num - 1], input[row_num + 1][col_num + 1]) && check_vals(input[row_num + 1][col_num - 1], input[row_num - 1][col_num + 1]);

}

int main() {
    std::ifstream file("input.txt");
    std::vector<std::string> input;
    std::string line;
    while (file >> line) {
        input.push_back(line);
    }
    int count = 0;
    for (size_t r = 1; r < input.size() - 1; ++r) {
        for (size_t c = 1; c < input[0].size() - 1; ++c) {
            count += check_coord(input, r, c);
        }
    }
    std::cout << count << std::endl;

    return 0;

}

File: day4/day4_2.py
from itertools import product

with open("input.txt") as file:
    array = file.readlines()

num_rows = len(array)
num_cols = len(array[0]) - 1 # -1 for \n at end of each line

def check_point(row, col):
    if array[row][col] != "A":
        return False
    diag1 = set([array[row-1][col-1], array[row+1][col+1]])
    diag2 = set([array[row-1][col+1], array[row+1][col-1]])
    return diag1 == diag2 == set("MS")
    

count = 0
for row_num in range(1, num_rows - 1):
    for col_num in range(1, num_cols - 1):
        count += check_point(row_num, col_num)
print(count)


